chain/chunks/src/lib.rs

```diff
@@ -123,6 +128,105 @@ impl RequestPool {
+}
+
+impl Seal {
+    fn process(&mut self, chunk_entry: &EncodedChunksCacheEntry) -> bool {
+        let mut part_ords_to_remove = HashSet::new();
+        let mut res = true;
+        for part_ord in self.part_ords.iter() {
+            if !chunk_entry.parts.contains_key(&part_ord) {
+                res = false;
+            } else {
+                part_ords_to_remove.insert(part_ord.clone());
+            }
+        }
+        for part_ord in part_ords_to_remove.drain() {
+            self.part_ords.remove(&part_ord);
+        }
+        res
+    }
+}
+
+pub struct SealsManager {
+    me: Option<AccountId>,
+    runtime_adapter: Arc<dyn RuntimeAdapter>,
+
+    seals: HashMap<ChunkHash, Seal>,
+    dont_include_chunks_from: HashSet<AccountId>,
+}
+
+impl SealsManager {
+    fn new(me: Option<AccountId>, runtime_adapter: Arc<dyn RuntimeAdapter>) -> Self {
+        Self {
+            me,
+            runtime_adapter,
+            seals: HashMap::new(),
+            dont_include_chunks_from: HashSet::new(),
+        }
+    }
+
+    fn get_seal(
+        &mut self,
+        chunk_hash: &ChunkHash,
+        parent_hash: &CryptoHash,
+        height: BlockHeight,
+        shard_id: ShardId,
+    ) -> Result<&mut Seal, Error> {
+        Ok(self.seals.entry(chunk_hash.clone()).or_insert({
+            let chunk_producer = self.runtime_adapter.get_chunk_producer(
+                &self.runtime_adapter.get_epoch_id_from_prev_block(parent_hash)?,
+                height,
+                shard_id,
+            )?;
+            let mut candidates = vec![];
+            for part_ord in 0..self.runtime_adapter.num_total_parts() {
+                let part_ord = part_ord as u64;
+                let part_owner = self.runtime_adapter.get_part_owner(parent_hash, part_ord)?;
+                if part_owner == chunk_producer || Some(part_owner) == self.me {
+                    continue;
+                }
+                candidates.push(part_ord);
+            }
+            let chosen = candidates
+                .choose_multiple(
+                    &mut rand::thread_rng(),
+                    cmp::min(NUM_PARTS_REQUESTED_IN_SEAL, candidates.len()),
+                )
+                .cloned()
+                .collect::<HashSet<_>>();
+            Seal { part_ords: chosen, chunk_producer, sent: Utc::now() }
+        }))
+    }
+
+    fn approve_chunk(&mut self, chunk_hash: &ChunkHash) {
+        let seal = self.seals.get_mut(chunk_hash).expect("seal should be already produced");
+        seal.part_ords.clear();
+    }
+
+    fn track_seals(&mut self) {
+        let now = Utc::now();
+        for (chunk_hash, seal) in self.seals.iter_mut() {
+            if seal.part_ords.len() > NUM_PARTS_LEFT_IN_SEAL
+                && (now - seal.sent).num_milliseconds() > ACCEPTING_SEAL_PERIOD_MS
+            {
+                warn!(target: "client", "Resented by {:?} for chunk grieving for chunk {:?}, I'm {:?}", seal.chunk_producer, chunk_hash, self.me);
```

Comment:
`Resented by {:?} for chunk grieving` -> `Couldn't reconstruct chunk from {:?}`

Reply:
++